CS4850Group5A/VirtuDoc
# Heroku Deploy

## Summary

- Added base Java Spring v2.6.2 template.
- Added Heroku deployment manifests.
- Added automatic Github Actions testing on PRs and pushes to `main`.
- Added CodeQL security scanning for PRs.
- Added Docker/Docker Compose environment that matches Heroku production environment.
- Added documentation on setting up local dev environment.

## How to Test
*Don't feel like you have to do all of these, but these steps are just all the ways to verify every change made by this PR.*
- Follow the instructions in `README.md` or in the [wiki](https://github.com/CS4850Group5A/VirtuDoc/wiki/Installing-Docker-and-Docker-Compose) to launch the entire application stack as a Docker Compose environment.
- Run `docker build --build-arg PORT=8080 -f prod.Dockerfile -t virtudoc-web .` and verify that the production image builds correctly. Then run `docker run -p 8080:8081 --env PORT=8081 virtudoc-web` and verify that the web application runs by ensuring that a test message appears when you visit `localhost:8080` in your browser.
- Run `./mvnw test` to confirm that unit tests run locally. (on Windows use `./mvnw.cmd test`)
- Check that Github Actions shows the results of the JUnit test suite in a comment below.
- When this pull request is merged, after about 5 minutes you should see a test message at [https://virtudoc.herokuapp.com/](https://virtudoc.herokuapp.com/).

## Additional Comments
### Assumptions made by this PR
This PR puts quite a few things in stone, that we should probably agree on before we start writing real code. Feel free if you want to discuss any of the below points through a comment or change request on this PR. Once we start building this app, these things will become increasingly difficult to change.
- Using Java 11 LTS (alternatives are 8, 11, and 17, versions with LTS support are supported best by Spring).
- Using Spring Boot v2.6.2 (can use any version number, but this is the most recent stable version).
- Using the AdoptJDK distribution of Java (only real alternative is HotSpot).
- Using Debian Linux as the host OS in the dev and production environment (only applies when using Docker, alternative is Alpine Linux).
- Using Maven for dependency management and `make`-style tasking (alternative is Gradle).
- Heroku will handle the final production image building, and not GitHub actions (alternative is to build on GHA and push to Heroku registry, and not benefit from configuration-as-code that the Heroku manifest offers).

### Impact on front-end
This PR does not make any assumptions about the design and implementation of the frontend, and can be discussed outside of this PR.

### Justification for Docker Compose
This PR also introduces Docker Compose as the all-in-one environment for running the application locally. Listed below are my reasons why I feel this extra tool is necessary:
- Docker Compose eliminates the need to install, configure, run, and manage database software locally.
- With Docker Compose, there is no trace left on your computer when you shut down the stack.
- Hardcoded Docker images ensures that everyone is running the exactly the same database/JDK/OS/auxiliary services, and running versions that match the ones running on Heroku in production.
- Developing on a running Docker instance eliminates environment errors that cannot be reproduced by everyone on the team.
- We are already using Docker images when deploying to Heroku, so much of the Docker configuration is still required anyways.
- Docker Compose only requires one command to run. There is no learning curve unless you want to edit the Docker stuff.
- The Docker stack is a requirement for #8 to work correctly, which may impact frontend development.
- IntelliJ and VS Code both have built-in tools for working inside a Docker Compose stack-based application, so the command-line instructions aren't even needed if you use one of those IDEs.
- Docker is a proven industry standard platform, and working with it is a nice way to pad your resume.

I feel that the above benefits outweigh the minor inconvenience of having to install one extra tool. I also feel that the one alternative that is Kubernetes is complete overkill, and would introduce more problems than it would solve. Of course, you are free to develop using the JDK installed on your host OS, and instructions are included in the PR on how to do so, but you will have to manually manage your own database services if you go that route. Please voice your concerns and possible other solutions below.

Closes #3
